src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ILScanNodeFactory.cs

```diff
@@ -40,6 +40,12 @@ protected override IMethodNode CreateMethodEntrypointNode(MethodDesc method)
 
             if (CompilationModuleGroup.ContainsMethodBody(method, false))
             {
+                // If this is an instance method on a non-valuetype, don't add direct reference to
+                // the method body just yet. We might be able to optimize the method body
+                // away if the owning type was never seen as allocated.
+                if (!method.Signature.IsStatic && !method.OwningType.IsValueType && !method.IsCanonicalMethod(CanonicalFormKind.Any))
```

Comment:
We should be able to enable this on canonical methods too, but it's a couple dozens of lines of code so I scoped it out for now to keep the pull request short.